Structured Concurrency

Aidan63 opened the pull request:
Some initial work for #32, this introduces a new `runScoped` function which accepts a coroutine which an initial function of the `CoroutineScope` type which allows spawning more coroutines. E.g. you can now do this.

```haxe
import haxe.coro.Coroutine;
import haxe.coro.Coroutine.delay;

function main() {
	Coroutine.runScoped(scope -> {
		scope.start(_ -> {
			delay(500);

			trace('Hello, ');
		});

		scope.start(_ -> {
			delay(1000);

			trace('World!');
		});

		trace('tasks launched!');
	});
}
```

Which prints out "tasks launched!", "Hello, ", and then "World!" in that order and the `runScoped` call won't return until all child coroutines have completed. I think this is the first time we've had multiple coroutines running on a single thread, which is quite nice.

The `start` function returns a created `Job` object, this class contains a suspending `await` function which can be used to manually wait for the job to complete. e.g.

```haxe
import haxe.coro.Coroutine;
import haxe.coro.Coroutine.delay;

function main() {
	Coroutine.runScoped(scope -> {
		scope.start(_ -> {
			delay(500);

			trace('Hello, ');
		});

		final job = scope.start(_ -> {
			delay(1000);

			trace('World!');
		});

		job.await();

		trace('tasks launched!');
	});
}
```

The first argument of that `start` function is another `CoroutineScope` so you can create even more coroutines in a sub scope. e.g. the following contrived example.

```haxe
import haxe.coro.Coroutine;
import haxe.coro.Coroutine.delay;

function main() {
    Coroutine.runScoped(scope -> {
        scope.start(_ -> {
            delay(500);

            trace('Hello, ');
        });

        final job = scope.start(inner -> {
            delay(1000);

            trace('first delay done');

            inner.start(_ -> {
                trace('inner delay starting...');

                delay(1000);

                trace('inner delay done');
            });

            trace('World!');
        });

        job.await();

        trace('tasks launched!');
    });
}
```

I've just been throwing this together so the implementation will need a fair bit of cleanup, there are also several things missing / needing to be done.

- None of this is thread safe, but really I think most of this should be. Targets which support threading are definitely going to want to run coroutines on different threads, so the internals definitely need to be robust against cross thread access and all the race conditions which may occur.
- The error propogation stuff is not implemented. If a coroutine errors, I don't know what happens!
- Somewhat related to the above, cancellation of jobs is also not implemented.
- I think I'm going to change `Job` and `CoroutineScope` to interfaces so the base continuation can implement them. I was already starting to run into inheritance issues with just this.

Simn commented:
Regarding thread-safety, I've been wondering if we should have stub implementations for non-thread targets in the standard library. It's very annoying to deal with this thread vs. non-thread situation, so having implementations of at least Mutex across all targets would greatly simplify this mess.

Aidan63 commented:
It's a tricky one, having a bunch of stub threading primitives on non threaded targets is weird and will make debugging / stepping through more confusing (not that any of it should surface to the user facing coroutine api). But the alternative (Foo.x.hx files which are 99% the same bar some missing locks) seems much worse, so it's probably the best solution for now.

Simn commented:
I have merged the Mutex addition so we can use that now.

> Returning a value from a job is not currently implented.

What does "from a job" mean here exactly?

Aidan63 commented:
A new coroutine started in a scope, i.e. the `await` function doesn't return a value so you can't do something like this.

```haxe
Coroutine.runScoped(scope -> {
	final job = scope.start(_ -> {
		yield();

		return 'Hello, World!';
	});

	final result = job.await();

	trace(result);
});
```

Kotlin has `launch` which returns a `Job` (can't be manually awaited as it returns unit), and `async` which returns `Defered<T>` and can be awaited for its return value (iirc there are also some slight differences in how they handles errors).
I was also thinking of moving away from the "Job" terminology, kotlin seems to use it since they have non coroutine jobs and I don't think that applies to us (yet anyway).

Aidan63 commented:
I did plan on doing some more work on this yesterday evening but got distracted making sample programs, so here's a another more complex one.
Eight coroutine producers all reading lines from files and writing them into a channel with a single reader then writing those lines into an output file. I made pretty much everything here async (including file io by bouncing it through a thread pool) so the execution of all nine coroutines are interleaved. I did get an error when trying to suspend in a catch block out of it, issue incoming...

```haxe
import sys.thread.FixedThreadPool;
import haxe.Exception;
import haxe.coro.IContinuation;
import haxe.coro.Coroutine;
import haxe.coro.Coroutine.delay;
import haxe.coro.Coroutine.yield;

final pool = new FixedThreadPool(4);

function main() {
    Coroutine.runScoped(scope -> {
        final output  = FileWriter.open('out.txt');
        final channel = new Channel();

        var count = 8;
        for (i in 0...8) {
            scope.start(_ -> {
                final input = FileReader.open('bin/lorem$i.txt');

                while (input.eof() == false) {
                    channel.write(input.readLine());
                }

                input.close();

                if (0 == --count) {
                    channel.close();
                }
            });
        }

        try {
            while (true) {
                output.writeLine(channel.read());
            }
        } catch (_:ChannelClosedException) {}

        output.close();
    });
}

class FileReader {
    final handle : sys.io.FileInput;

    @:coroutine public static function open(path : String) : FileReader {
        return Coroutine.suspend(cont -> {
            pool.run(() -> {
                cont.resume(new FileReader(sys.io.File.read(path)), null);
            });
        });
    }

    function new(handle) {
        this.handle = handle;
    }

    @:coroutine public function eof() : Bool {
        return Coroutine.suspend(cont -> {
            pool.run(() -> {
                cont.resume(handle.eof(), null);
            });
        });
    }

    @:coroutine public function readLine() : String {
        return Coroutine.suspend(cont -> {
            pool.run(() -> {
                cont.resume(handle.readLine(), null);
            });
        });
    }

    @:coroutine public function close() {
        return Coroutine.suspend(cont -> {
            pool.run(() -> {
                handle.close();

                cont.resume(null, null);
            });
        });
    }
}

class FileWriter {
    final handle : sys.io.FileOutput;

    @:coroutine public static function open(path : String) : FileWriter {
        return Coroutine.suspend(cont -> {
            pool.run(() -> {
                cont.resume(new FileWriter(sys.io.File.write(path)), null);
            });
        });
    }

    function new(handle) {
        this.handle = handle;
    }

    @:coroutine public function writeLine(line : String) {
        return Coroutine.suspend(cont -> {
            pool.run(() -> {
                handle.writeString('$line\n');

                cont.resume(null, null);
            });
        });
    }

    @:coroutine public function close() {
        return Coroutine.suspend(cont -> {
            pool.run(() -> {
                handle.close();

                cont.resume(null, null);
            });
        });
    }
}

class ChannelClosedException extends Exception {
    public function new() {
        super("Channel closed");
    }
}

class Channel<T> {
    final maxQueueSize = 3;
    final writeQueue = new Array<T>();
    final suspendedWrites = new Array<()->Void>();
    final suspendedReads = new Array<IContinuation<T>>();
    var closed = false;

    public function new() {}

    @:coroutine public function write(v:T):Void {
        if (closed) {
            throw new ChannelClosedException();
        }

        switch suspendedReads.shift() {
            case null:
                if (writeQueue.length < maxQueueSize) {
                    writeQueue.push(v);
                } else {
                    Coroutine.suspend(cont -> {
                        suspendedWrites.push(() -> {
                            writeQueue.push(v);

                            cont.resume(null, null);
                        });
                    });
                }
            case cont:
                cont.resume(v, null);
        }
    }

    @:coroutine public function read():T {
        if (writeQueue.length < maxQueueSize && suspendedWrites.length > 0) {
            suspendedWrites.shift()();
        }

        switch writeQueue.shift() {
            case null:
                if (closed) {
                    throw new ChannelClosedException();
                }

                return Coroutine.suspend(cont -> {
                    suspendedReads.push(cont);
                });

            case v:
                return v;
        }
    }

    public function close() {
        closed = true;

        while (suspendedReads.length > 0) {
            suspendedReads.shift().resume(null, new ChannelClosedException());
        }
    }
}
```

Simn commented:
> I was also thinking of moving away from the "Job" terminology, kotlin seems to use it since they have non coroutine jobs and I don't think that applies to use (yet anyway).

I agree with this, `Job` seems quite odd in all this. The relationship between jobs and contexts is still not very clear to me, especially because in Kotlin both form a tree structure from what I understand.

Simn commented:
Question: would it be crazy to implement `Coroutine.run` as `return runScoped(_ -> f());` to avoid all this code duplication?

Aidan63 commented:
I was thinking of going one further, will we even need `Coroutine.run` at the end of this? Do we just bin that and rename `runScoped` to `run`, if you're only writing a trivial coroutine and don't care about the scope then just `_` it out.

Simn commented:
Yes, if you put it like that it seems even better.

Aidan63 commented:
I've done some code cleanup and moved away from the `Job` terminology to (the perhaps equally confusing) `ICoroutine`. In the same way that the `IContinuation` interface forms the execution stack of a given coroutine this `ICoroutine` interface exposes access to parent and child coroutines, so you get a tree of coroutines. The coroutine context now has a field to access the `ICoroutine` of the current coroutine...
Stuff like that is why I think this name is still a bit confusing, but I don't know what else to call it since it does represent the overall coroutine. I think stuff like this is might be part of why kotlin chose the `suspend` keyword instead of `coroutine` for functions.

The new `BaseCoroutine` class implements several interfaces and is currently the main class of interest for all the internals. I've also updated `await` to return a value now so the following is possible.

```haxe
function main() {
    Coroutine.runScoped(scope -> {
        final c = scope.start(_ -> {
            delay(500);

            return 'Hello, World!';
        });

        final result = c.await();

        trace(result);
    });
}
```

Error propagation is still one of the major things to do, that will require tackling cancellation of coroutines since all sibling coroutines need to be cancelled in the case of an error before it's passed up to the parent. Porting over the cancellation token stuff from my haxe-coroutines fork is probably a good starting point.

The other is then thread safety which should now be do-able with the threading primitive stubs, easiest way to test that for now would probably be a `startWith` function on `ICoroutineScope` which allows specifying a scheduler for the new coroutines context.

Simn commented:
> Stuff like that is why I think this name is still a bit confusing, but I don't know what else to call it since it does represent the overall coroutine. I think stuff like this is might be party of why kotlin chose the `suspend` keyword instead of `coroutine` for functions.

I've reached that conclusion as well. However, even in Kotlin this terminology is not used consistently. The [mega document](https://github.com/JetBrains/kotlin/blob/master/compiler/backend/src/org/jetbrains/kotlin/codegen/coroutines/coroutines-codegen.md) goes into this too:

> Throughout the document term "coroutine" will represent either a suspend lambda or a suspend function, which is different from the usual definition of coroutines - something like a lightweight thread. The document reuses the term since "suspend lambda or function" is wordy, and when it requires the typical definition, it says explicitly "a coroutine in a broad sense."

Simn commented:
I just tried running your file reader sample. At first all it would do is hang, which I guess is because of errors that go unnoticed if the `bin/lorem0.txt` files don't exist. After creating these the program terminates, but my out.txt has the content of lorem7.txt eight times:

```
7
7
7
7
7
7
7
7
```

Not sure if something is wrong with the sample logic, the coro logic or my setup.

Simn commented:
I struggle to describe the problem in words, but we basically have this situation:

```haxe
function main() {
	var k = 8;
	var a = [];
	var hoisted = { i: 0 };
	for (i in 0...k) {
		hoisted.i = i;
		a.push(() -> trace(hoisted.i));
	}
	for (f in a) {
		f();
	}
}
```

So the stored function accesses `hoisted.i`, which by the time it executes has already been incremented all the way.

Aidan63 commented:
I created a bunch of those lorem files which contain lines such as

```
lorem0 line 1
lorem0 line 2
lorem0 line 3
```

and so on, then in the out.txt file you see those individual lines interleaved (due to all the suspension)

```
lorem1 line 1
lorem0 line 1
lorem2 line 1
lorem1 line 2
lorem0 line 2
lorem1 line 3
lorem0 line 3
lorem1 line 4
```

I was testing it with cpp and jvm and it seemed to work fine, I was not seeing the 7th file repeated.

Aidan63 commented:
Ah, a hoisting issue. Could that sort of thing be why kotlin saves and restores local variables instead of remapping all access like we do?

Simn commented:
I can't see how that would help, if the hoisted value is wrong by the time we execute then spilling it into a local isn't going to change that.

But I'm also quite confused why this doesn't reproduce for you...

Aidan63 commented:
This is pretty silly. Initially when I made that demo I went to a lorem ipsum generator site, generated some paragraphs, bunged it in a file, and then copy and pasted that 8 times. It appeared to work! The lines were interleaved, but because it was the same paragraphs in all the files I never noticed that issue.

However... I since deleted my bin folder so this morning when I was looking at it again I created just three lorem files with a handful of those `lorem0 line 0` lines since I didn't want to go though all that faff again. Three is an important number here, it seems thats small enough to where the compiler will unroll that loop, so then it does actually work...

Aidan63 commented:
So I guess a small sample would be something like this.

```haxe
function main() {
    Coroutine.runScoped(scope -> {
        for (i in 0...10 + Std.random(10)) {
            yield();

            scope.start(_ -> {
                trace('$i');
            });
        }
    });
}
```

This sample is a bit nastier since it does sometimes actually work, but sometimes you'll get a single number appear twice due to scheduler timing issues introduced by that `yield` call.
The kotlin style variable saving and restoring would solve this one though, right? When yield is called it would save the value of `i` and then when it resumes it would restore a local variable with that saved value. The `scope.start` lambda would then capture that local variable so will always have the correct value.

Simn commented:
This one fails consistently for me:

```haxe
function main() {
	Coroutine.runScoped(scope -> {
		for (i in 0...10 + Std.random(10)) {
			yield();
			scope.start(_ -> {
				delay(10);
				trace('$i');
			});
		}
	});
}
```

And yes, I _think_ detouring over a spilled local would help here, but thinking about this also gives me a headache...

Aidan63 commented:
The following sample fails to compile on cpp (works on js and jvm).

```haxe
import haxe.Exception;
import haxe.coro.IContinuation;
import haxe.coro.Coroutine;
import haxe.coro.Coroutine.delay;
import haxe.coro.Coroutine.yield;

function main() {
    Coroutine.runScoped(scope -> {
        scope.start(_ -> {
            yield();
            
            trace('hello');
        });
    });
}
```

The error is a C++ error about `_g` not being defined.

```
Compiling group: haxe
g++ -Iinclude -c -fvisibility=hidden -g -fpic -fPIC -Wno-overflow -DHX_LINUX -DHXCPP_DEBUG -DHXCPP_M64 -DHXCPP_VISIT_ALLOCS(haxe) -DHX_SMART_STRINGS(haxe) -DHXCPP_API_LEVEL=500(haxe) -m64 -DHXCPP_M64 -I/mnt/d/programming/haxe/hxcpp/include ... tags=[haxe,static]
 - src/_Main/Main_Fields_.cpp 
Error: ./src/_Main/Main_Fields_.cpp: In static member function ‘static void _Main::Main_Fields__obj::main()’:
./src/_Main/Main_Fields_.cpp:189:111: error: ‘_g’ was not declared in this scope
  189 | HXDLIN(   8)            ::haxe::coro::_Coroutine::Coroutine_Impl__obj::runScoped( ::Dynamic(new _hx_Closure_0(_g)));
      |                                                                                                               ^~
 ERROR  (unknown position)

   | Error: Build failed
```

Looking at a pretty dump it seems like `_g` is defined within a block but is then accessed outside.

```haxe
private class _Main.Main_Fields_ {

	@:keep
	public static function main() {
		haxe.coro._Coroutine.Coroutine_Impl_.runScoped(@:mergeBlock cast function(scope:haxe.coro.ICoroutineScope, _hx_completion:haxe.coro.IContinuation<Any>) {
			try {
				{
					var _g = [];
					_g[0] = null;
					_g[0] = function(_:haxe.coro.ICoroutineScope, _hx_completion:haxe.coro.IContinuation<Any>) {
						// state machine omitted
					};
					_g[0] = _g[0];
				};
				return haxe.coro.ImmediateSuspensionResult.withResult(scope.start(@:mergeBlock _g[0]));
			} catch (_g1:Dynamic) {
				{
					null;
				};
				return haxe.coro.ImmediateSuspensionResult.withError(_g1);
			};
			return null;
		});
	}
}
```

Edit: Adding a yield to the closure passed to `runScoped` gets it to compile, so it seems to be in cases where no suspension actually occurs.

Simn commented:
I'll take a look. Could you merge kt_coro into this PR?

Simn commented on std/haxe/coro/coroutines/ChildCoroutine.hx:
```diff
@@ -0,0 +1,14 @@
+		final ctx    = parentContext.clone();
+		final parent = parentContext.get(AbstractCoroutine.key);
+
+		super(ctx, (cast parent : ICoroutine<Any>));
```
I don't know if I'm supposed to pay attention to such things yet, but this cast looks quite suspicious. This tells me that either AbstractCoroutine should already implement ICoroutine or that we want to work with BaseCoroutine instead.

Aidan63 replied:
This is a dodgy cast and I mainly put it here to get things working with the merge faster. Tricky thing here is that `BaseCoroutine` is generic so would it then have a unique key for each type parameter version? Getting the coroutine is then also odd, should you do `ctx.get(BaseCoroutine<SomeActualType>.key)` or `ctx.get(BaseCoroutine<Any>.key)`, neither seem very good.

So in the end I added the non generic `AbstractCoroutine` but that does then require a cast to get the coroutine interface.

Simn replied:
I don't think type parameters should be relevant here because in the general case we don't know the exact type of our parent coroutine anyway. So just `ctx.get(BaseCoroutine.key)` is what I would expect, with `class BaseCoroutine<T> extends Element<BaseCoroutine<Any>>`.

Although this immediately makes me wonder if Element should be an interface instead with a `getKey()` function...

Aidan63 replied:
I think the intermittent auto completion was throwing me off then, I didn't think I could do `BaseCoroutine.key`. I've made `Coroutine` have a key but given it the type `ICoroutine<Any>` so you can now do `ctx.get(Coroutine.key)` to get the `ICoroutine`.

Simn commented on std/haxe/coro/Coroutine.hx:
```diff
@@ -22,10 +24,12 @@ private class CoroSuspend<T> extends haxe.coro.BaseContinuation<T> {
 @:callable
 @:coreType
 abstract Coroutine<T:haxe.Constraints.Function> {
+	public static final key : Key<ICoroutine<Any>> = Key.createNew('_hx_coroutine');
```
This gave me pause because Coroutine itself isn't actually part of any of the hierarchies here, but on a nominal level it makes a lot of sense to have this here.

Aidan63 replied:
I've been trying to avoid exposing the heirarchy stuff anywhere to give us some leeway if we need to change it down the line, putting that key on `Coroutine` seemed to make sense.

Aidan63 commented:
I've implemented some basic child to parent error handling, right now it only works with scopes which have a single child. So with the following very nested sample

```haxe
import haxe.Exception;
import haxe.coro.IContinuation;
import haxe.coro.Coroutine;
import haxe.coro.Coroutine.delay;
import haxe.coro.Coroutine.yield;

@:coroutine function baz() {
	throw new Exception('hello');
}

@:coroutine function bar() {
	baz();
}

@:coroutine function foo() {
	bar();
}

function main() {
	Coroutine.runScoped(scope -> {
		scope.start(scope -> {
			scope.start(_ -> {
				foo();
			});
		});
	});
}
```

you get a stack trace looking like this.

```
Called from _Main.Main_Fields_.baz (Main.hx line 7)
Called from _Main.Main_Fields_.bar (Main.hx line 11)
Called from _Main.Main_Fields_.foo (Main.hx line 15)
Called from haxe.coro.coroutines.BaseCoroutine.start (haxe/coro/coroutines/BaseCoroutine.hx line 69)
Called from sys.thread.EventLoop.progress (sys/thread/EventLoop.hx line 146)
Called from haxe.coro._EventLoop.EventLoop_Impl_.tick (haxe/coro/EventLoop.hx line 17)
Called from haxe.coro.coroutines.BlockingCoroutine.wait (haxe/coro/coroutines/BlockingCoroutine.hx line 22)
Called from haxe.coro._Coroutine.Coroutine_Impl_.runScoped (haxe/coro/Coroutine.hx line 79)
Called from _Main.Main_Fields_.main (Main.hx line 20)
Called from haxe.coro.coroutines.BlockingCoroutine.wait (haxe/coro/coroutines/BlockingCoroutine.hx line 41)
Called from haxe.coro._Coroutine.Coroutine_Impl_.runScoped (haxe/coro/Coroutine.hx line 79)
Called from _Main.Main_Fields_.main (Main.hx line 20)
Error : hello
```

There is something odd going on with the stack traces though, if you add a `yield` call in `bar` just before the `baz` call, most of the call stack goes missing.

```
Called from _Main.Main_Fields_.baz (Main.hx line 7)
Called from _Main.Main_Fields_.bar (Main.hx line 13)
Called from haxe.coro.coroutines.BlockingCoroutine.wait (haxe/coro/coroutines/BlockingCoroutine.hx line 41)
Called from haxe.coro._Coroutine.Coroutine_Impl_.runScoped (haxe/coro/Coroutine.hx line 79)
Called from _Main.Main_Fields_.main (Main.hx line 21)
Error : hello
```

Simn commented:
Multiple problems:

1. The original stack trace is already wrong because it points to the function definition, not the actual call. This must be related to some inaccurate position.
2. There's a problem with TCO again because `ImmediateSuspensionResult` just provides `[]` as its stack, which is then taken over by the calling coroutine (`bar` here). The whole mechanism expects the coroutine that actually causes the exception to also set up the stack, so this is something we'll have to check.
3. The entire `completeExceptionally` thing ignores the `result`, which holds the stack trace in the case of an error.

Simn commented:
There's even a second problem with TCO which is obvious if we look at how `foo` is generated:

```java
    public static SuspensionResult<SuspensionResult<Unit>> foo(IContinuation<java.lang.Object> _hx_completion) {
        try {
            return bar(_hx_completion);
        } catch (Throwable var2) {
            return (SuspensionResult)ImmediateSuspensionResult.withError(Exception.caught(var2).unwrap());
        }
    }
```

It simply passes through `_hx_completion`, so our `IStackFrame` mechanism has no place to put an additional entry to make it part of the coro stack. It's also not part of the top or bottom stacks, so it ends up being skipped completely.

I'm starting to think that we'll have to go back to the drawing board for our stack trace handling. It seemed like a good idea to pass this information via `result`, but it looks like this causes quite a few problems down the road.

Simn commented:
I've made it pass through the `result` value in all places that seem relevant. With this your example now _almost_ gives the correct stacktrace:

```
Exception in thread "main" hello
        at _Main.Main_Fields_.baz(source/Main.hx:8)
        at _Main.Main_Fields_.bar(source/Main.hx:13)
        at _Main.Main_Fields_.bar(source/Main.hx:13)
        at _Main.Main_Fields_.foo(source/Main.hx:17)
        at _Main.Main_Fields_$Closure_main_0$Closure_invoke_0$Closure_invoke_0.invoke(source/Main.hx:24)
        at haxe.coro.coroutines.BlockingCoroutine.wait(C:\git\haxe\std/haxe/coro/coroutines/BlockingCoroutine.hx:41)
        at haxe.coro.Coroutine$Coroutine_Impl_.runScoped(C:\git\haxe\std/haxe/coro/Coroutine.hx:79)
        at _Main.Main_Fields_.main(source/Main.hx:22)
        at _Main.Main_Fields_.main(source/Main.hx)
```

I don't know why line 13 shows up twice. It doesn't happen with a normal `Coroutine.run(foo)` so there must be some detail I'm overlooking again.

Aidan63 commented:
what sample do you get that duplicated item with? I'm not seeing it. I'm also still seeing the cut off stack trace if I put yield before the baz call from that previous sample, but I may have broke something with the additions I just pushed.

Speaking of breaking things I just tried to run that foo bar baz sample on jvm and get a class cast exception, seems related to the stack building in `BlockingCoroutine`.

```
Exception in thread "main" java.lang.ClassCastException: class haxe.coro.coroutines.ChildCoroutine cannot be cast to class haxe.root.Array (haxe.coro.coroutines.ChildCoroutine and haxe.root.Array are in unnamed module of loader 'app')
        at haxe.coro.coroutines.BlockingCoroutine.wait(/usr/local/share/haxe/std/haxe/coro/coroutines/BlockingCoroutine.hx:40)
        at haxe.coro.Coroutine$Coroutine_Impl_.runScoped(/usr/local/share/haxe/std/haxe/coro/Coroutine.hx:79)
        at _Main.Main_Fields_.main(Main.hx:25)
        at _Main.Main_Fields_.main(Main.hx)
```

Simn commented:
https://github.com/Aidan63/haxe/pull/66/commits/b4583431409ff4560a32c929b72cc7d5091210c2 inherited `completed.error` but not `completed.result`.

The basic rule is this: **if we assign error, also assign result because that's where the stack trace is expected to be**.

We probably want to make this fail nicer in the future, but for now it's good that this explodes so that we notice where it's missing.

I've also fixed the duplicate entry problem. I forgot that the stack handling was copied to BlockingCoroutine, so the logic changes in BlockingContinuation were never applied.

Aidan63 commented:
Updates on this branch.

I think `runScoped` and `start` is pretty much how it should be working now which is good. E.g. when a scope throws all child coroutines are issued cancellation requests and waited for their completion.

```haxe
Coroutine.runScoped(scope -> {
	scope.start(_ -> {
		delay(1000);
	});

	throw new Exception("");
});
```

In that example `runScoped` will wait until the child has exited even though the outer coroutine has errored (eventually `delay` will be more prompt in responding to cancellations, but it shows the conncept nicely right now). There is also some initial way to poll for cancellation.

```haxe
Coroutine.runScoped(scope -> {
	scope.start(_ -> {
		while (scope.context.get(Coroutine.key).isCancelled == false) {
			yield();
		}
	});

	delay(1000);

	throw new Exception("");
});
```

It is important to note that with this structured concurrency catching the exception thrown from the `await` will still cause the parent to receive cancellation.

```haxe
// runScoped will still throw an exception even though we have a try catch around await.
Coroutine.runScoped(scope -> {
	final child = scope.start(_ -> {
		delay(1000);

		throw new Exception("");
	});

	try {
		child.await();
	} catch (_:Exception) {
		//
	}
});
```

The new `Coroutine.scope` suspending function allows creating a scope which will return a value or raise an exception and not lead to the immediate cancellation of the parent.

```haxe
Coroutine.runScoped(scope -> {
	try {
		Coroutine.scope(scope -> {
			scope.start(_ -> {
				while (scope.context.get(Coroutine.key).isCancelled == false) {
					yield();
				}
			});

			delay(1000);

			throw new Exception("");
		});
	} catch (_:Exception) {
		//
	}
});
```

There are a number of things which still need to be done.

- Coroutine.scope is still a bit buggy, need to track them down.
- Cancellation of a parent does not propagate down to `Coroutine.scope`
- Still needs to be thread safe. I think we might be able to "cheat" this a bit by scheduling the execution of most public functions onto the coroutines scheduler. Only the handful of properties would then need to be protected by a mutex.
- Need a way to allow users to provide custom context elements.
- Want to add something similar to kotlin's `SupervisorJob` which then allows custom error handling.
- I'm going to add a `haxe.exceptions.CancellationException` and have that be treated as a special case internally. This was we can differenciate between a coroutine failing exceptionally or failing due to cancellation. In the later a coroutine doesn't need to tell it's parent it's errored since it's parent will have been the one to issue the cancellation.
- Take another pass at the `BaseCoroutine` internals, I'm not too happy with how some of it has grown.
- Nicer user facing API for interacting with cancellations (polling and receiving a notification of cancellation).

Simn commented:
Thank you for the update!

My initial reaction to the runScoped example that cancels the parent upon a child exception is that it violates the principle of least surprise. The code _looks_ like it should catch the exception, but it doesn't. I understand why that is the case (I think), but I don't really like it... I'll play around with it a bit and hopefully get a better picture.

Simn commented on std/haxe/coro/coroutines/ChildCoroutine.hx:
```diff
@@ -0,0 +1,11 @@
+
+import haxe.coro.context.Context;
+
+class ChildCoroutine<T> extends BaseCoroutine<T> {
```
These days I'm getting pretty paranoid about classes and the `extends` keyword, so let me confirm this: is it really the case that a coroutine can be either a ChildCoroutine or a ScopeCoroutine? The distinction isn't very clear to me, so I'd like to make sure.

Aidan63 replied:
I'm not sure yet, there is still work to be done on `ScopeCoroutine` so maybe they'll be rolled into one once done. There's also the question of a supervisor coroutine as well.

Simn replied:
Just going by names, my intuition would be that something can be both a child coroutine and a (scope coroutine or a supervisor coroutine). Although it's hard to tell because both ChildCoroutine and ScopeCoroutine don't have any distinct functionality yet, so it's more like a tag right now.

Aidan63 commented:
I agree that the exception being raised to the parent seems odd even though you have a try catch around `await`. It might not be very idiomatic coroutine code though, so maybe it won't come up that much in the real world.
Kotlin also has a few functions which work well with that sample. First is `coroutineScope` (`Coroutine.scope` for us) which raises the final error of that scope through the suspending function continuation instead of going to the parent. The other is the `supervisorScope` where children created within it don't cause cancellation of siblings, so you can `await` a child to deal with it's error.

Simn commented:
> Take another pass at the BaseCoroutine internals, I'm not too happy with how some of it has grown.

About this, I've been thinking that we should focus on composition more. I had this shower thought that what we're doing here with coroutines is kind of similar to an entity-component-system: the coroutine is the entity and all the stuff we add to its context are the components, which are then operated on by various systems. I don't know how relevant this is for the actual implementation, but it helps me conceptualize everything better.

Aidan63 commented:
Yes, I think that makes sense. Most of `BaseCoroutine` is managing a state machine for the coroutine, might make sense to have that class be just state machine management and expose a series of callbacks for child failed, etc, then ChildCoroutine, ScopeCoroutine, and whatever else can weave those together how they see fit.

Simn commented:
For educational purposes I tried implementing a component-based approach on this branch: https://github.com/Aidan63/haxe/commit/d3c0bb0b4133764d29de89a9477f12519a990c18

Instead of having all the business logic on BaseCoroutine itself, it now lives in the three components I've added:

* CancelComponent: This is created at the "entry point" (BlockingCoroutine) and inherited by all contexts. I don't know how useful having cancellation as a component could really be, but I can think of scenarios where we _really_ don't want to have something cancelled and end up overriding it with a `NullCancelComponent` or something along those lines.
* ChildrenComponent: This holds the children data and is copied when creating a child context. This one might be pretty useless, but I like the separation of concerns, so I'm not sure! (Edit: after thinking about this some more, I think this one is silly and should be removed again.)
* ScopeComponent: This is the interesting one. I've copied the logic from the `start` function to `DefaultScopeComponent`, but the idea here is that we could just replace this in the context with something like `SupervisorScopeComponent`, which I think is what we want.

At the moment they all live in BaseCoroutine.hx because that's easier to manage. I imagine they would be moved to a haxe.coro.components package if we decide to do this.

As always, I don't know if this makes any sense.

Aidan63 commented:
I like the idea, one thing I've also wanted to do is make it so that creating children isn't part of the `BaseCoroutine`. Kotlin has it's `start` equivalent be a static extension on the scope, would be nice to have something similar here as it "proves" it's not hard coded so then it's "hopefully" easy for users to provide their own custom functionality. Having the structured concurrency behaviour be composeable with component seems to align with that.

Kotlin has a `NonCancellable` scope which it recommends for stuff like finally blocks which want to call async functions, so there will definitely be a use for some sort of `NullCancellableComponent`.

Simn commented:
The problem with static extensions is that we need a place to store the actual data, which in that case could only be the context again. But this might actually be a good approach indeed because it felt wrong to have the `cancel` function on `BaseCoroutine` for convenience, when all the other cancellation logic has been outsourced to the component.

The advantage would be that this truly decouples the logic from the coroutine. The disadvantage is that this wouldn't work through Dynamic and such. Hmm...

Simn commented:
I've added `DefaultScopeComponent`, which is `CancelComponent` and `DefaultScopeComponent` fused together because I think it makes sense for this to be one unit. I'm pretty confident in that part.

The next step is to actually make the context adjustable which we still can't do. I've made a new [demo commit](https://github.com/Aidan63/haxe/commit/9a7ab6b8c89d880840767dab7bdb12be7b80f116) which picks up my idea from [here](https://github.com/Aidan63/haxe/pull/78#issuecomment-2882642753) to give us this kind of approach:

```haxe
class TracingScopeComponent extends ScopeComponent {
	public function new() {}

	public function cancel(coroutine:BaseCoroutine<Any>) {
		trace(coroutine);
	}

	public function onCompletion(coroutine:BaseCoroutine<Any>, childCoroutine:BaseCoroutine<Any>) {
		trace(coroutine, childCoroutine);
	}
}

function main() {
	Coroutine.runScoped(scope -> {
		var coro = scope.with(new TracingScopeComponent()).start(scope -> {
			trace(scope);
		});
		coro.cancel();
		coro.await();
	});
}
```

I like this `with` approach because it's very flexible and even allows to spawn multiple coroutines with the same adjusted context. This example does trace in `cancel`, but I don't see the one in `onCompletion`. I'm not quite sure if I should see that one.

Edit: Also I find this `Component` suffix strangely reassuring because I can mentally group everything related to that together without having to think about what everything is. I wouldn't be opposed to using this on all related concepts, i.e. the schedulers and such.

Aidan63 commented:
I've just checked out that branch and I think you are missing some sort of `onCompletion` callback. E.g. the following sample never exits. you never see the "exiting child" trace but do see the "cancelled" one, so I think cancellation is not being pass in correctly

```haxe
function main() {
	Coroutine.runScoped(scope -> {
		final child = scope.start(scope -> {
			while (scope.context.get(Coroutine.key).isCancelled == false) {
				yield();
			}

			trace('exiting child');
		});
		
		delay(1000);

		child.cancel();

		trace('cancelled');
	});
}
```

The good news is that the branch does work a bit better with the commented out test I added on this branch, but it still doesn't completely work.

```haxe
// on the "structured" branch you never see the "exiting child scope" trace, but you do on "structured_with_with"
// it does hang forever on both branches though.
function main() {
	Coroutine.runScoped(scope -> {

		final child = scope.start(_ -> {

			Coroutine.scope(scope -> {
				while (scope.context.get(Coroutine.key).isCancelled == false) {
					yield();
				}

				trace('exiting coroutine scope');
			});

			trace('exiting child scope');
		});
		
		delay(1000);

		child.cancel();

		trace('cancelled');
	});
}
```

I'll take a look tomorrow morning and see if I can figure out whats going on (in either branch), but I'm going to be away from mid day tomorrow to sunday so if I don't figure it out tomorrow I won't be looking at it again for a few days.

Simn commented:
Hmm, that commit wasn't supposed to change any of the logic, and I can't see why it would. If something is broken is should be broken on this branch here as well, so if anything the culprit should be https://github.com/Aidan63/haxe/pull/66/commits/e385f9037ca717d602c257b3e3b3fd8a2ec66499

Aidan63 commented:
Seems like the `BaseCoroutine` constructor was no longer setting `this` on the context (think it use to be done in `ChildCoroutine` ctor), so the parent child structure was not being established correctly.

It is a bit worrying that none of our existing tests caught this... makes me wonder if we're missing something or some abstraction here isn't quite right.

Aidan63 commented:
thinking about this I think it probably wasn't spotted because we don't have the things in place yet for what kotlin calls "prompt cancellation guarentee". E.g. if a cancellation occurs in a middle of a delay it is woke up early instead of waiting for the delay to end, or cancellation is checked for before resuming from `Coroutine.suspend`.

Simn commented:
Ah good spot! I had dismissed `ChildCoroutine` as "the thing that doesn't do anything" so I didn't look at it properly. But always setting it in the `BaseCoroutine` constructor might be more robust anyway.

In fact, it should probably be there where we also call `clone` to avoid situations where two coroutines share the same context.

The reason this didn't cause any problems is because nobody does `context.get(Coroutine.key)`. At least the test from https://github.com/Aidan63/haxe/pull/66/commits/dca9de2b4e5b891012af49908b6911124de79364 does that now.

Note that my example from https://github.com/Aidan63/haxe/pull/66#issuecomment-2897653225 still doesn't trace in `onCompletion`. The code looks like it should do that, or am I missing something?

Simn commented:
On another note, can we reduce the delays in the tests a bit? Waiting for these is a little grating and always makes me think something is broken.

Simn commented on std/haxe/coro/scopes/DefaultScopeComponent.hx:
```diff
@@ -0,0 +1,66 @@
+import haxe.coro.coroutines.BaseCoroutine;
+
+@:access(haxe.coro.coroutines.BaseCoroutine)
+class DefaultScopeComponent extends ScopeComponent {
```
I wonder if stateless components like this should be turned into a singleton. There's no reason to allocate them for each coroutine. The only question is where to put this and when/how to initialize it.

Simn commented:
Regarding your hanging scope problem, I noticed that the traces aren't consistent. When running your sample I sometimes see this:

```
source/Main.hx:42: cancelled
source/Main.hx:32: exiting coroutine scope
source/Main.hx:35: exiting child scope  
```

And sometimes this:

```
source/Main.hx:42: cancelled
source/Main.hx:32: exiting coroutine scope
```

Looks like some kind of race condition.

Edit: Adding a trace to the `child.onCompletion` callback in `Coroutine.scope` shows that we have `state == Completed` and `error == Cancellation exception`. Is that something that is supposed to happen? I'd expect the state to be `Cancelled` here.

Edit: Well, I suppose if we let a coroutine complete and then cancel it this would be the outcome, so maybe that's alright.

Aidan63 commented:
On delays, yes I think they can be reduced. I want to get around to #76 sooner rather than later so there are no "real" delays in the tests.

Simn commented:
The most recent commit addresses a problem with how we were handling callbacks because I realized that a callback could cause the array we're iterating over to be modified, which is undefined behavior.

Your sample terminates now, albeit probably not how we want it to:

```
source/Main.hx:40: cancelled
source/Main.hx:30: exiting coroutine scope
Exception in thread "main" Cancellation exception
        at haxe.coro.scopes.DefaultScopeComponent.cancel(C:\git\haxe\std/haxe/coro/scopes/DefaultScopeComponent.hx:13)
        at haxe.coro.coroutines.BaseCoroutine.cancel(C:\git\haxe\std/haxe/coro/coroutines/BaseCoroutine.hx:153)
        at haxe.coro.coroutines.BaseCoroutine.completeExceptionally(C:\git\haxe\std/haxe/coro/coroutines/BaseCoroutine.hx:198)
        at haxe.coro.scopes.DefaultScopeComponent.cancel(C:\git\haxe\std/haxe/coro/scopes/DefaultScopeComponent.hx:13)
        at haxe.coro.coroutines.BaseCoroutine.cancel(C:\git\haxe\std/haxe/coro/coroutines/BaseCoroutine.hx:153)
```

Edit: this comes from the `cont.resume(null, child.error)` call in `Coroutine.scope`, which is turned into a real exception by `runScoped`. Should we just call it with `null` instead? Also, shouldn't the `Completed` state pass `child.error` as second argument?

Edit: I've decided that we should because it makes the disabled test pass!

Simn commented:
Hmm, there's some conflation here regarding cancellation and exceptions. I thought the function named `completeExceptionally` would actually set the state to `Completed`, but it is `Cancelled` instead. This is something we need to sort out because we don't want to leak these CancellationExceptions, which makes it necessary to distinguish these cases in `Coroutine.scope` at the very least.

Edit: Could you give me the job description of `Coroutine.scope` because that one is still quite elusive.

Aidan63 commented:
`Coroutine.scope` allows you to create a child scope right where you are, without needing to explicitly call `.start` on some scope reference. It follows all the same structured concurrency semantics but with a small difference in how it surfaces any errors.
Errors from `Coroutine.scope` are raised as standard exceptions and the parent isn't implicitly cancelled. This means you can try catch around it to supress / handle any errors unlike `await`.

Aidan63 commented:
For an extra point of reference, the kotlin equivilent below just prints out "cancelled" and the process exits with error code 0.

```kotlin
fun main() {
    runBlocking {
        val child = async {
            coroutineScope {
                while (isActive) {
                    yield()
                }

                println("exiting coroutine scope")
            }

            println("exiting child scope")
        }

        delay(1000)

        child.cancel()

        println("cancelled")
    }
}
```

I'm now just trying to convince my self that this makes sense and that I understand it...

Simn commented:
Does that actually run the child at all though? I see `async` so I expect `await` or something else that starts the actual execution.

Aidan63 commented:
`async` isn't lazy by default, so it should have started. For paranoia's sake I added a `child.start` call which returns false, docs say that means it didn't start as a result of that call (already started). Putting an extra `println` before the while loop starts does show in the console.

Aidan63 commented:
I added a `child.await` call in there and now it does throw a `JobCancellationException`, I'm even more confused... I thought errors from `async` were propagated up to their parent. This does cause an exception from the top level with or without the `await`.

```kotlin
fun main() {
    runBlocking {
        val child = async {
            coroutineScope {
                throw Exception("hello")
            }
        }

        delay(1000)

        child.await()
    }
}
```

So I guess this comes back to the comment you made earlier about needing to handle cancellation exceptions differently?

Simn commented:
I guess that depends on what we want. From a usage standpoint I expect a `child.cancel()` call to cancel the child and leave me alone otherwise.

Aidan63 commented:
Ok, so I think I get whats going on now. Any child which completes exceptionally with a `CancellationException` shouldn't then cancel on it's parent (it does currently) which makes sense since the parent will have been one to cancel it.

For us thats currently all done though the `onCompletion` callbacks (which I've never been a fan of). I wonder instead if `ICoroutine` should have a `parent` field of type `ICoroutine` and then if a child completes with something other than a `CancellationException` it then calls `parent.cancel(error)` and we keep the `onCompletion` callbacks for book keeping.

Aidan63 commented:
On a somewhat related note it would be nice to have `ICoroutine` expose the parent and children somehow, this would allow walking the coroutine tree in the same way you can walk the continuation heirarchy. Could allow generating graphs for the current coroutine structure which will probably be useful, would also open up debuggers to implement something like full fat visual studios parallel stacks view by walking that tree.

https://learn.microsoft.com/en-us/visualstudio/debugger/using-the-parallel-stacks-window?view=vs-2022#tasks-view

Simn commented:
> For us thats currently all done though the `onCompletion` callbacks (which I've never been a fan of). I wonder instead if `ICoroutine` should have a `parent` field of type `ICoroutine` and then if a child completes with something other than a `CancellationException` it then calls `parent.cancel(error)` and we keep the `onCompletion` callbacks for book keeping.

That sounds good to me, although it seems too narrow: when a child completes (in whatever way) it should notify its parent, which can then react depending on the child's (and its siblings) state.

Simn commented:
Here are some observations after looking at everything again and basically reimplementing most of it to better understand the architecture:

## Exposing states

This is a nominal mess at the moment:

* In `CoroutineState`, there are `Completed` and `Cancelled` which are clearly disjunctive, but then there's `isCompleted` which is true for both of them. 
* `isRunning` is only true for `Running` itself, which gives the false impression that the coroutine isn't doing anything in the cases of  `Completing` and `Cancelling`.
* `isCancelled` is also true for `Cancelling`, which gives the false impression that the coroutine is cancelled even though it's still processing the cancellation.

I don't know if this is a "when in Rome" situation which we inherited from Kotlin, but I don't like it either way. We should design this in a way that makes sense for a *public* interface, which is unlikely to care about implementation details like `Cancelling` and `Completing`. To me, the following questions make sense from an outside perspective:

* *Is the coroutine still doing something?* (`Running | Completing | Cancelling`) This should be `isRunning`.
* *Can the coroutine be cancelled?* (`Running | Completing`) This should be `isCancellable`.
* *Did the coroutine complete successfully?* (`Completed`)
* *Did the coroutine complete unsuccessfully?* (`Cancelled`)

I'm thinking that we shouldn't directly expose the latter two because that makes it easy to make the following mistake:

```haxe
if (coro.completedSuccessully()) {
    // good
} else {
    // oh no, it must have failed!
}
```

But this isn't true because it could also still be running. This makes me think that we should expose a separate tri-state enum here which is similar to `SuspensionState`.

## await and scope

I can't shake off the feeling that the way `await` works with regards to exceptions is an annoying limitation and prompted the introduction of `Coroutine.scope` as basically a workaround. I still maintain that this should work:

```haxe
try {
    coro.await();
} catch(e:Dynamic) {}
```

The fact that it doesn't suggests to me that something is wrong with how we implemented `await`. Surely if we await a specific child and then that very child errors, our control flow should return to the point where we're awaiting. This suggests to me that there should be an interaction between `await` and `resume`, where resuming with an error while we're awaiting should defer further execution to the await continuation.

I'm not sure how to actually implement that though.

## Cancellation propagation

This one gives me a bit of a headache. I think we generally agree that exceptions and cancellations aren't the same thing: while exceptions should cause the cancellation of all children and bubble up to the parent (in the default scope anyway), cancellations shouldn't be so eager to be propagated to the parent.

Intuitively, I'd say that if we have multiple children and cancel one of them, then its siblings should not be affected by default. If we depend on the result of said child, we'll likely do so via `await`, and in that case the cancellation should be propagated automatically. I think.

## Parent relationships

I agree that coroutines should know their parents and notify them of their completion. This allows the parent to take stock of how its children are doing and manage its state accordingly, much like what we're doing with the completion callbacks at the moment.

In my new implementation (which is still way too messy even for me to push somewhere), I'm handling this via a new interface that is implemented by my equivalent of `BaseCoroutine`:

```haxe
interface ICoroutineHost {
	function childCompletes<T>(child:ICoroutine<T>, result:T):Void;
	function childErrors(child:ICoroutine<Any>, error:Exception):Void;
	function childCancels(child:ICoroutine<Any>, error:Exception):Void;
}
```

And then all the places which ultimately set the state to `Completed` or `Cancelled` call these like `parent?.childCompletes(this, result);`. My `childErrors` implementation is this:

```haxe
	public function childErrors(child:ICoroutine<Any>, error:Exception) {
		completedChildren++;
		context.get(ScopeComponent.key).childErrors(this, child, error);
		checkCompletion();
	}
```

This is where we defer the concrete handling to the scope. The `DefaultScopeComponent` will cancel the parent, while the `SupervisorScopeComponent` will not do anything. I think that is what we want.

## Closing remarks

I'll continue playing with all this and see if I can get the branch to a state that I can push. My biggest problem is still the `await` function, so we'll see how that goes.

Simn commented:
> `Coroutine.scope` allows you to create a child scope right where you are, without needing to explicitly call `.start` on some scope reference. It follows all the same structured concurrency semantics but with a small difference in how it surfaces any errors.
> Errors from `Coroutine.scope` are raised as standard exceptions and the parent isn't implicitly cancelled. This means you can try catch around it to supress / handle any errors unlike `await`.

Coming back to this, are you sure this is the intended specification? I checked the [docs of the Kotlin version](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/coroutine-scope.html) and it states this quite clearly:

> When any child coroutine in this scope fails, this scope fails, cancelling all the other children (for a different behavior, see [supervisorScope](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/supervisor-scope.html)).

To me this sounds like it should have the exact same behavior as `start` in this regard.

The reason it works differently right now is because it calls `child`, which has the `completedChildren++` onCompletion handling, but it doesn't call `startChild` which has `coroutine.onCompletion(() -> context.get(ScopeComponent.key).onCompletion(this, coroutine))`, which is what causes the exception behavior normally. This is quite subtle and I can't tell if it's working as intended or if it's more of an accident.

Aidan63 commented:
I think I mangled my words a bit in that comment, I'm in agreement with the kotlin docs (I think). E.g. in this sample the longer delaying task in the scope never runs to completion and try catch around `coroutineScope` catches the exception which means it isn't automatically propagated up to the `runBlocking` which remains uncancelled.

```kotlin
fun main() {
    runBlocking {

        async {
            delay(1000)
            println("ran to completion")
        }

        try {
            coroutineScope {
                async {
                    delay(500)
                    throw ArithmeticException()
                }

                async {
                    delay(1000)
                    println("done")
                }
            }
        }
        catch (exn:ArithmeticException) {
            println("caught exception")
        }
    }
}
```

Simn commented:
Hmm... but if child errors cause their parents to error as well, how come the error from the `coroutineScope` block doesn't do that for its own parent? I must still be mixing up the concepts of job/scope and cancellation/exception somehow.

Simn commented:
I think I just had a revelation by looking at how Java handles all this. We've been making this harder on ourselves by conflating scopes and jobs in our `BaseCoroutine`, which might be the root cause of all my confusion.

In Java there's an [StructuredTaskScope](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/StructuredTaskScope.html) class which has a `fork` method to spawn `Subtask` objects. That method is basically the equivalent of Kotlin's `async` and our `start`, but the trick here is that the scope is shared by all created subtasks. This actually seems totally logical if I spell it out like that, but it's not what we have implemented so far: Since our `BaseCoroutine` implements both `ICoroutine` and `ICoroutineScope`, we effectively create a new scope for every child.

A failing child causes everything in the scope to be cancelled, and the difference with `coroutineScope` is that it creates an entirely new scope which can fail by itself without failing its caller. I'm not saying "parent" because this explicit relationship doesn't actually seem to exist between scopes. A scope will simply do its thing and then potentially raise an exception, which can be caught normally from the outside.

Now I'm inclined to try an API that is closer to Java's with the `fork` and `join` methods, but still using Kotlin's context approach because I really like that. But either way, I feel like there's hope again!

Simn commented:
Superseded by #97.
